### PR TITLE
refactor: use stateful openshift client instead of ad hoc (#664)

### LINKED ISSUE #664: OpenShift client is not-reused, it is re-created everytime it's needed
https://github.com/redhat-developer/intellij-openshift-connector/blob/4a7c11e208923bf13106ec4eb2f79923dcdfed6c/src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java#L589

https://github.com/redhat-developer/intellij-openshift-connector/blob/4a7c11e208923bf13106ec4eb2f79923dcdfed6c/src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java#L834-L837

https://github.com/redhat-developer/intellij-openshift-connector/blob/4a7c11e208923bf13106ec4eb2f79923dcdfed6c/src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java#L860-L861

`ClusterHelper.isOpenShift(client)` is also creating a new OpenShift client:
https://github.com/redhat-developer/intellij-common/blob/9c0aae5b4917000596cd192f2ba150c33ca893e5/src/main/java/com/redhat/devtools/intellij/common/kubernetes/ClusterHelper.java#L28
```java
  public static boolean isOpenShift(KubernetesClient client) {
        OpenShiftClient osClient = (OpenShiftClient)client.adapt(OpenShiftClient.class);
        ...
```

This is especially time consuming when it first asks whether the cluster is an OpenShift cluster (which creates an OpenShift client) and then creates an OpenShift client to retrieve data.

https://github.com/redhat-developer/intellij-openshift-connector/blob/4a7c11e208923bf13106ec4eb2f79923dcdfed6c/src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java#L227-L229

### PR DESCRIPTION
**Pull request by adietish**
fixes #664

### CODE REVIEW

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 1]
```diff
@@ -225,14 +228,11 @@ public List<String> getNamespaces() throws IOException {
     }
 
     private List<? extends HasMetadata> getNamespaces(KubernetesClient client) {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:230]: there is no need to use this param as it is the kubernetesclient instance. please remove and use directly the `client ` field

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 2]
```diff
@@ -153,6 +155,7 @@ public OdoCli(com.intellij.openapi.project.Project project, String command) {
         String current = ConfigHelper.getCurrentContextName();
         Config config = Config.autoConfigure(current);
         this.client = new KubernetesClientBuilder().withConfig(config).withHttpClientBuilderConsumer(builder -> setSslContext(builder, config)).build();
+        this.openshiftClient = toOpenShiftClient(client);
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:158]: wonder if using the client as param makes sense ? `toOpenShiftClient` method is used only at this place now.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 3]
```diff
@@ -373,11 +373,11 @@ public List<ComponentMetadata> analyze(String path) throws IOException {
     public void createComponent(String project, String componentType, String registryName, String component, String source, String devfile, String starter) throws IOException {
         List<String> args = new ArrayList<>();
         args.add("init");
-        if (StringUtils.isNotBlank(devfile)) {
+        if (!Strings.isEmptyOrSpaces(devfile)) {
             args.add("--devfile-path");
             args.add(devfile);
         } else {
-            if (StringUtils.isNotBlank(starter)) {
+            if (!Strings.isEmptyOrSpaces(starter)) {
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:380]: please do same refactor in line 655

Comment (sbouchet) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:380]: and 895

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 4]
```diff
@@ -833,15 +832,12 @@ public java.net.URL getMasterUrl() {
...
-                    VersionInfo info = oclient.getOpenShiftV3Version();
-                    if (info == null) {
-                        ConfigMap configMap = client.configMaps().inNamespace(OCP4_CONFIG_NAMESPACE).withName(OCP4_CONSOLE_PUBLIC_CONFIG_MAP_NAME).get();
-                        if (configMap != null) {
-                            return configMap.getData().get(OCP4_CONSOLE_URL_KEY_NAME);
-                        }
+            if (openshiftClient != null) {
+                VersionInfo info = openshiftClient.getOpenShiftV3Version();
+                if (info == null) {
+                    ConfigMap configMap = client.configMaps().inNamespace(OCP4_CONFIG_NAMESPACE).withName(OCP4_CONSOLE_PUBLIC_CONFIG_MAP_NAME).get();
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:838]: just tested out that this code needs to use openshiftClient versus client. tested out that it actually results the same, and will improve readability.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 5]
```diff
@@ -833,15 +832,12 @@ public java.net.URL getMasterUrl() {
...
+                    if (configMap != null) {
+                        return configMap.getData().get(OCP4_CONSOLE_URL_KEY_NAME);
                     }
                 } else {
                     ConfigMap configMap = client.configMaps().inNamespace(OCP3_CONFIG_NAMESPACE).withName(OCP3_WEBCONSOLE_CONFIG_MAP_NAME).get();
```
Comment (sbouchet) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:843]: same as above

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 2 above)
Comment (adietish) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java:158]: I would keep it. Method signature lays bare that a client is adapted, something you'd only discover reading the method body if the instance variable was used. Furthermore it makes it immediately clear that one needs to initialize the kubernetes client first.

**Review by sbouchet — APPROVED** [review]
LGTM

**Pull request merged**